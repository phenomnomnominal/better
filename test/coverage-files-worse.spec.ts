import { createFixture } from './fixture';

describe('betterer', () => {
  it('should report a worse file coverage result', async () => {
    const { betterer } = await import('@betterer/betterer');

    const fixture = await createFixture('coverage-files-worse', {
      '.betterer.js': `
const { coverage } = require('@betterer/coverage');

module.exports = {
  test: () => coverage()
};    
        `,
      'tsconfig.json': `
{
  "extends": "../../tsconfig.json",
  "include": ["./src/**/*", "./.betterer.js"]
}
      `
    });

    const file1Path = fixture.resolve('./src/file-1.ts');
    const file2Path = fixture.resolve('./src/file-2.ts');
    const file3Path = fixture.resolve('./src/file-3.ts');

    const coverage = `{  
  "total": {
    "lines": { "total": 220, "covered": 110, "skipped": 0, "pct": 50 },
    "statements": { "total": 410, "covered": 140, "skipped": 0, "pct": 34.1 },
    "functions": { "total": 30, "covered": 11, "skipped": 0, "pct": 36.6 },
    "branches": { "total": 300, "covered": 100, "skipped": 0, "pct": 33.3 }
  },
  "${file1Path}": {
    "lines": { "total": 100, "covered": 50, "skipped": 0, "pct": 50 },
    "statements": { "total": 200, "covered": 70, "skipped": 0, "pct": 35 },
    "functions": { "total": 20, "covered": 5, "skipped": 0, "pct": 25 },
    "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
  },
  "${file2Path}": {
    "lines": { "total": 120, "covered": 60, "skipped": 0, "pct": 50 },
    "statements": { "total": 210, "covered": 70, "skipped": 0, "pct": 33.3 },
    "functions": { "total": 10, "covered": 6, "skipped": 0, "pct": 60 },
    "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
  }
}`;

    const configPaths = [fixture.paths.config];
    const resultsPath = fixture.paths.results;
    const coveragePath = fixture.resolve('./coverage/coverage-summary.json');

    await fixture.writeFile(coveragePath, coverage);

    await betterer({ configPaths, resultsPath, workers: false });

    // file-1 is worse
    const worseCoverage = `{  
      "total": {
        "lines": { "total": 220, "covered": 100, "skipped": 0, "pct": 45.4 },
        "statements": { "total": 410, "covered": 140, "skipped": 0, "pct": 34.1 },
        "functions": { "total": 30, "covered": 11, "skipped": 0, "pct": 36.6 },
        "branches": { "total": 300, "covered": 100, "skipped": 0, "pct": 33.3 }
      },
      "${file1Path}": {
        "lines": { "total": 100, "covered": 40, "skipped": 0, "pct": 40 },
        "statements": { "total": 200, "covered": 70, "skipped": 0, "pct": 35 },
        "functions": { "total": 20, "covered": 5, "skipped": 0, "pct": 25 },
        "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
      },
      "${file2Path}": {
        "lines": { "total": 120, "covered": 60, "skipped": 0, "pct": 50 },
        "statements": { "total": 210, "covered": 70, "skipped": 0, "pct": 33.3 },
        "functions": { "total": 10, "covered": 6, "skipped": 0, "pct": 60 },
        "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
      }
    }`;

    await fixture.writeFile(coveragePath, worseCoverage);

    const worseTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(fixture.testNames(worseTestRun.worse)).toEqual(['test']);

    // file-1 is better, but file-2 is worse
    const stillWorseCoverage = `{  
      "total": {
        "lines": { "total": 220, "covered": 90, "skipped": 0, "pct": 40.9 },
        "statements": { "total": 410, "covered": 140, "skipped": 0, "pct": 34.1 },
        "functions": { "total": 30, "covered": 11, "skipped": 0, "pct": 36.6 },
        "branches": { "total": 300, "covered": 100, "skipped": 0, "pct": 33.3 }
      },
      "${file1Path}": {
        "lines": { "total": 100, "covered": 60, "skipped": 0, "pct": 60 },
        "statements": { "total": 200, "covered": 70, "skipped": 0, "pct": 35 },
        "functions": { "total": 20, "covered": 5, "skipped": 0, "pct": 25 },
        "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
      },
      "${file2Path}": {
        "lines": { "total": 120, "covered": 30, "skipped": 0, "pct": 25 },
        "statements": { "total": 210, "covered": 70, "skipped": 0, "pct": 33.3 },
        "functions": { "total": 10, "covered": 6, "skipped": 0, "pct": 60 },
        "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
      }
    }`;

    await fixture.writeFile(coveragePath, stillWorseCoverage);

    const stillWorseTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(fixture.testNames(stillWorseTestRun.worse)).toEqual(['test']);

    // file-3 is added but it is not 100% covered
    const evenWorseCoverage = `{  
      "total": {
        "lines": { "total": 320, "covered": 150, "skipped": 0, "pct": 46.8 },
        "statements": { "total": 610, "covered": 210, "skipped": 0, "pct": 34.4 },
        "functions": { "total": 50, "covered": 16, "skipped": 0, "pct": 32 },
        "branches": { "total": 450, "covered": 150, "skipped": 0, "pct": 33.3 }
      },
      "${file1Path}": {
        "lines": { "total": 100, "covered": 60, "skipped": 0, "pct": 60 },
        "statements": { "total": 200, "covered": 70, "skipped": 0, "pct": 35 },
        "functions": { "total": 20, "covered": 5, "skipped": 0, "pct": 25 },
        "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
      },
      "${file2Path}": {
        "lines": { "total": 120, "covered": 30, "skipped": 0, "pct": 25 },
        "statements": { "total": 210, "covered": 70, "skipped": 0, "pct": 33.3 },
        "functions": { "total": 10, "covered": 6, "skipped": 0, "pct": 60 },
        "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
      },
      "${file3Path}": {
        "lines": { "total": 100, "covered": 60, "skipped": 0, "pct": 60 },
        "statements": { "total": 200, "covered": 70, "skipped": 0, "pct": 35 },
        "functions": { "total": 20, "covered": 5, "skipped": 0, "pct": 25 },
        "branches": { "total": 150, "covered": 50, "skipped": 0, "pct": 33.3 }
      }
    }`;

    await fixture.writeFile(coveragePath, evenWorseCoverage);

    const evenWorseTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(fixture.testNames(evenWorseTestRun.worse)).toEqual(['test']);

    expect(fixture.logs).toMatchSnapshot();

    await fixture.cleanup();
  });
});
