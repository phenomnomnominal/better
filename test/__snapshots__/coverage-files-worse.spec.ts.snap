// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`betterer should report a worse file coverage result 1`] = `
Array [
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ test: \\"test\\" got checked for the first time! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got checked for the first time! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got checked for the first time! 🎉

1 test got checked. 🤔
1 test got checked for the first time! 🎉
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"lines\\" coverage is worse in \\"src/file-1.ts\\": 60 > 50

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"lines\\" coverage is worse in \\"src/file-1.ts\\": 60 > 50

Error: \\"test\\" got worse. 😔

1 test got checked. 🤔
1 test got worse. 😔

You should try to fix the new issues! As a last resort, you can run \`betterer --update\` to force an update of the results file. 🆙
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"lines\\" coverage is better in \\"src/file-1.ts\\": 40 < 50
・ \\"lines\\" coverage is worse in \\"src/file-2.ts\\": 90 > 60

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"lines\\" coverage is better in \\"src/file-1.ts\\": 40 < 50
・ \\"lines\\" coverage is worse in \\"src/file-2.ts\\": 90 > 60

Error: \\"test\\" got worse. 😔

1 test got checked. 🤔
1 test got worse. 😔

You should try to fix the new issues! As a last resort, you can run \`betterer --update\` to force an update of the results file. 🆙
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"lines\\" coverage is better in \\"src/file-1.ts\\": 40 < 50
・ \\"lines\\" coverage is worse in \\"src/file-2.ts\\": 90 > 60
・ new file: src/file-3.ts

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"lines\\" coverage is better in \\"src/file-1.ts\\": 40 < 50
・ \\"lines\\" coverage is worse in \\"src/file-2.ts\\": 90 > 60
・ new file: src/file-3.ts

Error: \\"test\\" got worse. 😔

1 test got checked. 🤔
1 test got worse. 😔

You should try to fix the new issues! As a last resort, you can run \`betterer --update\` to force an update of the results file. 🆙
",
]
`;
